### PR TITLE
Luigi cookbook

### PR DESCRIPTION
**Pull request by gmendonca**
Creating luigi cookbook.

### CODE REVIEW

File: templates/default/luigi-server.init.erb [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+PIDFILE=<%= @luigid_pidfile %>
+LOGDIR=<%= @luigid_logdir %>
+DAEMON_ARGS="--background --pidfile $PIDFILE --logdir $LOGDIR"
+SCRIPTNAME=/etc/init.d/$NAME
+
+# Exit if the package is not installed
+[ -x "$DAEMON" ] || exit 0
+
+# Read configuration variable file if it is present
+[ -r /etc/default/$NAME ] && . /etc/default/$NAME
```
Comment (barata) [inline templates/default/luigi-server.init.erb:32]: Does it make sense to create `luigi-env.sh.erb` in `/etc/default/luigid`?
If so, we could define attributes like `default['luigid']['env']['VALUE']` to be include in the template file.

File: templates/default/client.cfg.erb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[scheduler]
+record_task_history = True
+state_path = <%= @luigid_statefile %>
+
+[task_history]
+db_connection = sqlite:///<%= @luigi_sqlite %>
```
Comment (barata) [inline templates/default/client.cfg.erb]: What if we want to support JDBC? Is it possible?

File: templates/default/client.cfg.erb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+[scheduler]
+record_task_history = True
```
Comment (barata) [inline templates/default/client.cfg.erb]: Always `True`? Can we use lowercase?

File: templates/default/client.cfg.erb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+[scheduler]
```
Comment (barata) [inline templates/default/client.cfg.erb]: Is there other properties besides these? Can we generalize this file?

File: recipes/default.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+end
+
+template "etc/init.d/luigid" do
+  source    'luigi-server.init.erb'
+  owner     'root'
+  group     'root'
+  mode      '0755'
+  action    :create
+  notifies  :enable, 'service[luigid]'
+  notifies  :start, 'service[luigid]'
```
Comment (barata) [inline recipes/default.rb]: `notifies  :start, 'service[luigid]' if node['luigid']['auto_start']`

File: recipes/default.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
+#
+# Cookbook Name:: luigi
+# Recipe:: default
+#
+
+include_recipe 'python'
+
+directory node['luigi']['dir'] do
```
Comment (barata) [inline recipes/default.rb]: What if `node['luigi']['user']` or `node['luigi']['group']` does not exist?

File: attributes/default.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+default['luigi']['user'] = 'luigi'
+default['luigi']['group'] = 'luigi'
+default['luigi']['dir'] = '/etc/luigi'
+
+default['luigid']['user'] = 'luigi'
+default['luigid']['group'] = 'luigi'
+default['luigid']['dir'] = '/mnt/luigid'
```
Comment (barata) [inline attributes/default.rb]: This structure sounds strange. It is not clear what goes inside `default['luigid']['dir']`

**Review by barata — CHANGES_REQUESTED** [review]

File: templates/default/client.cfg.erb (hunk 2 above)
Comment (gmendonca) [reply templates/default/client.cfg.erb]: Couldn't find anything out of the configuration.

File: templates/default/client.cfg.erb (hunk 3 above)
Comment (gmendonca) [reply templates/default/client.cfg.erb]: Python uses `True` not `true`.

File: templates/default/luigi-server.init.erb [hunk 8]
```diff
@@ -0,0 +1,158 @@
+#! /bin/sh
```
Comment (mwconceicao) [inline templates/default/luigi-server.init.erb:1]: Consider creating systemd unit config for systems that moved to systemd (like Ubuntu 16.04).

**Comment by gmendonca** [comment]
@barata what do you think now?

File: attributes/default.rb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+default['luigi']['user'] = 'luigi'
+default['luigi']['group'] = 'luigi'
+default['luigi']['config_dir'] = '/etc/luigi'
+
+default['luigi_server']['auto_start'] = true
+default['luigi_server']['log_dir'] = '/usr/local/var/log'
+default['luigi_server']['pid_file'] = '/usr/local/var/luigid.pid'
```
Comment (barata) [inline attributes/default.rb]: Usually we use `/var/run/luigid.pid`.

File: attributes/default.rb [hunk 10]
```diff
@@ -0,0 +1,16 @@
+default['luigi']['user'] = 'luigi'
+default['luigi']['group'] = 'luigi'
+default['luigi']['config_dir'] = '/etc/luigi'
+
+default['luigi_server']['auto_start'] = true
+default['luigi_server']['log_dir'] = '/usr/local/var/log'
```
Comment (barata) [inline attributes/default.rb]: Usually we use `/var/log/<app>`.

File: attributes/default.rb [hunk 11]
```diff
@@ -0,0 +1,16 @@
+default['luigi']['user'] = 'luigi'
+default['luigi']['group'] = 'luigi'
+default['luigi']['config_dir'] = '/etc/luigi'
+
+default['luigi_server']['auto_start'] = true
```
Comment (barata) [inline attributes/default.rb]: What about `default['luigi']['server']['auto_start']`?

File: attributes/default.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+default['luigi_server']['log_dir'] = '/usr/local/var/log'
+default['luigi_server']['pid_file'] = '/usr/local/var/luigid.pid'
+
+default['client_cfg']['scheduler'] = {
+  'record_task_history' = True, # Python syntaxe
```
Comment (barata) [inline attributes/default.rb]: This has nothing to do with python. This is ruby, and that's it :)

File: attributes/default.rb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+default['luigi']['user'] = 'luigi'
+default['luigi']['group'] = 'luigi'
+default['luigi']['config_dir'] = '/etc/luigi'
+
+default['luigi_server']['auto_start'] = true
+default['luigi_server']['log_dir'] = '/usr/local/var/log'
+default['luigi_server']['pid_file'] = '/usr/local/var/luigid.pid'
+
+default['client_cfg']['scheduler'] = {
```
Comment (barata) [inline attributes/default.rb]:
```
default['luigi'][client_cfg] = {
  'scheduler' => {
    'record_tasl_history' => true,
    'state_path' => '/usr/local/var/luigi-state.pickle'
  },
  'task_history' => {
    'db_connection' => 'sqlite:////usr/local/var/luigi-task-hist.db'
  }
}
```

File: recipes/default.rb [hunk 14]
```diff
@@ -0,0 +1,66 @@
+#
+# Cookbook Name:: luigi
+# Recipe:: default
+#
+
+include_recipe 'python'
```
Comment (barata) [inline recipes/default.rb]: Move this close to where it is used.

File: recipes/default.rb [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
```
Comment (barata) [inline recipes/default.rb]: Define an attribute `luigi.setup_user` and create one here if it is true.
(The same applies for group).

File: recipes/default.rb [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+directory node['luigi']['config_dir'] do
+  owner     node['luigi']['user']
+  group     node['luigi']['group']
+  mode      '0755'
+  action    :create
+end
+
+template "#{node['luigi']['config_dir'}/client.cfg" do
+  source    'client.cfg.erb'
+  owner     node['luigi']['user'] || 'root'
```
Comment (barata) [inline recipes/default.rb]: Do you really need this `||`?

File: recipes/default.rb [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+  group     node['luigi']['group'] || 'root'
+  mode      '0755'
+  action    :create
+  variables({
+    :client_cfg => node['client_cfg']
+  })
+end
+
+directory node['luigi_server']['log_dir'] do
+  owner     node['luigi']['user'] || 'root'
```
Comment (barata) [inline recipes/default.rb]: Do you need this `||`?

File: recipes/default.rb [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+  group     node['luigi']['group'] || 'root'
+  mode      '0755'
+  action    :create
+  variables({
+    :client_cfg => node['client_cfg']
+  })
+end
+
+directory node['luigi_server']['log_dir'] do
```
Comment (barata) [inline recipes/default.rb]: enable recursive dir creation.

File: recipes/default.rb [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+end
+
+python_pip 'boto'
+python_pip 'python-daemon'
+python_pip 'SQLAlchemy'
+python_pip 'luigi'
+
+service 'luigid' do
+  supports :restart => true, :start => true, :stop => true, :reload => true
+  action :nothing
```
Comment (barata) [inline recipes/default.rb:75]: `action :enable`

File: recipes/default.rb [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+end
+
+template "etc/init.d/luigid" do
+  source    'luigi-server.init.erb'
+  owner     node['luigi']['user'] || 'root'
+  group     node['luigi']['group'] || 'root'
+  mode      '0755'
+  action    :create
+  notifies  :enable, 'service[luigid]'
```
Comment (barata) [inline recipes/default.rb]: I think you can remove this.

File: recipes/default.rb [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+  notifies  :start, 'service[luigid]' if node['luigid']['auto_start']
```
Comment (barata) [inline recipes/default.rb]: `luigid` does not exist. :-(

File: templates/default/luigi-server.init.erb [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+#
+# Function that starts the daemon/service
+#
+do_start()
+{
+	# Return
+	#   0 if daemon has been started
+	#   1 if daemon was already running
+	#   2 if daemon could not be started
+	start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON --test > /dev/null \
```
Comment (barata) [inline templates/default/luigi-server.init.erb:51]: `--test`?

File: templates/default/luigi-server.init.erb [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+# Do NOT "set -e"
+
+# PATH should only include /usr/* if it runs after the mountnfs.sh script
+PATH=/sbin:/usr/sbin:/bin:/usr/bin:/usr/local/bin
+DESC="Luigi central planner server"
+NAME=luigid
+DAEMON=/usr/local/bin/$NAME
+PIDFILE=<%= @luigid_pidfile %>
+LOGDIR=<%= @luigid_logdir %>
+DAEMON_ARGS="--background --pidfile $PIDFILE --logdir $LOGDIR"
```
Comment (barata) [inline templates/default/luigi-server.init.erb:25]: `--pidfile` is duplicated. Please check which one is really necessary.

File: templates/default/luigi-server.init.erb [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+do_start()
+{
+	# Return
+	#   0 if daemon has been started
+	#   1 if daemon was already running
+	#   2 if daemon could not be started
+	start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON --test > /dev/null \
+		|| return 1
+	start-stop-daemon --start --quiet --pidfile $PIDFILE --exec $DAEMON -- \
+		$DAEMON_ARGS \
```
Comment (barata) [inline templates/default/luigi-server.init.erb:54]: `--pidfile` is also defined on `$DEAMON_ARGS`.

**Review by barata — CHANGES_REQUESTED** [review]
A lot of changes.

File: recipes/default.rb (hunk 19 above)
Comment (gmendonca) [reply recipes/default.rb:75]: I saw [this](http://stackoverflow.com/questions/9938314/chef-how-to-run-a-template-that-creates-a-init-d-script-before-the-service-is-c) thread, that's why I kept `:nothing` here.

Comment (barata) [reply recipes/default.rb:75]: If it doesn't work, put this `service` statement right after `template` creation below. But I'd first try it here.

File: recipes/default.rb [hunk 25]
```diff
@@ -2,8 +2,16 @@
...
 # Recipe:: default
 #
+user node['luigi']['user'] do
+  system  true
+  shell   "/bin/false"
+  only_if node['luigi']['setup_user']
+end
 
-include_recipe 'python'
+group node['luigi']['group'] do
```
Comment (barata) [inline recipes/default.rb]: You need to associate the user to this group.
First create the group, then create the user referencing this group.

**Comment by gmendonca** [comment]
@barata everything has been tested. I think we're all set here.

File: attributes/default.rb [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+default['luigi']['setup_group'] = true
+default['luigi']['user'] = 'luigi'
+default['luigi']['group'] = 'luigi'
+default['luigi']['config_dir'] = '/etc/luigi'
+
+default['luigi']['server']['auto_start'] = true
+default['luigi']['server']['log_dir'] = '/var/log/luigi'
+default['luigi']['server']['pid_file'] = '/var/run/luigid.pid'
+
+default['luigi']['server']['create_dir'] = true
```
Comment (barata) [inline attributes/default.rb]: I think this parameter is not necessary. If `luigi.server.dir` does not exist, we always create it, that's it.

File: recipes/default.rb [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+# Cookbook Name:: luigi
+# Recipe:: default
+#
+user node['luigi']['user'] do
+  system  true
+  shell   '/bin/false'
+  only_if { node['luigi']['setup_user'] }
+end
+
+group node['luigi']['group'] do
```
Comment (barata) [inline recipes/default.rb]: I think I've already said that. But again... you need to associate luigi user to luigi group.

File: recipes/default.rb [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+  system  true
+  only_if { node['luigi']['setup_group'] }
+end
+
+directory node['luigi']['config_dir'] do
```
Comment (barata) [inline recipes/default.rb]: `recursive`?

File: recipes/default.rb [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+  group     node['luigi']['group']
+  mode      '0755'
+  action    :create
+end
+
+template "#{node['luigi']['config_dir']}/client.cfg" do
+  source    'client.cfg.erb'
+  owner     node['luigi']['user']
+  group     node['luigi']['group']
+  mode      '0755'
```
Comment (barata) [inline recipes/default.rb]: 644? (rw, r, r)

Comment (barata) [reply recipes/default.rb]: directories need `x` permission, config files don't.

File: recipes/default.rb [hunk 30]
```diff
@@ -0,0 +1,85 @@
...
+  recursive true
+  action    :create
+end
+
+directory node['luigi']['server']['dir'] do
+  owner     node['luigi']['user']
+  group     node['luigi']['group']
+  mode      '0755'
+  action    :create
+  only_if { node['luigi']['server']['create_dir'] }
```
Comment (barata) [inline recipes/default.rb]: dont think we need this.

File: recipes/default.rb [hunk 31]
```diff
@@ -0,0 +1,85 @@
...
+
+directory node['luigi']['server']['log_dir'] do
+  owner     node['luigi']['user']
+  group     node['luigi']['group']
+  mode      '0755'
+  recursive true
+  action    :create
+end
+
+directory node['luigi']['server']['dir'] do
```
Comment (barata) [inline recipes/default.rb]: `recursive`?

File: recipes/default.rb [hunk 32]
```diff
@@ -0,0 +1,85 @@
...
+  owner     node['luigi']['user']
+  group     node['luigi']['group']
+  mode      '0755'
+  action    :create
+  only_if { node['luigi']['server']['create_dir'] }
+end
+
+template '/etc/default/luigid' do
+  source    'client.cfg.erb'
```
Comment (barata) [inline recipes/default.rb]: did you really test this? Oo

**Review by barata — CHANGES_REQUESTED** [review]

**Comment by gmendonca** [comment]
@barata

**Review by barata — APPROVED** [review]

**Pull request merged**